src/app/games/martian-chess/MartianChessMove.ts

```diff
@@ -0,0 +1,74 @@
+
+    public static readonly END_COORD_OUT_OF_RANGE: string = 'End coord cannot be out of range';
+
+    public static readonly PAWN_MUST_MOVE_ONE_DIAGONAL_STEP: () => string = () => $localize`Pawns must move one diagonal step`;
+
+    public static readonly DRONE_MUST_DO_TWO_ORTHOGONAL_STEP: () => string = () => $localize`Drones must move two orthogonal steps`;
+}
+
+export class MartianChessMove extends MoveCoordToCoord {
+
+    public static encoder: NumberEncoder<MartianChessMove> = NumberEncoder.tuple(
+        [Coord.numberEncoder(4, 8), Coord.numberEncoder(4, 8), NumberEncoder.booleanEncoder],
+        (move: MartianChessMove): [Coord, Coord, boolean] => [move.coord, move.end, move.calledTheClock],
+        (f: [Coord, Coord, boolean]): MartianChessMove => MartianChessMove.from(f[0], f[1], f[2]).get(),
+    ); // TODO FOR REVIEW: une fois qu'on a des examples, j'avoue ça défonce du ponichon !
+    public static from(start: Coord, end: Coord, calledTheClock: boolean = false): MGPFallible<MartianChessMove> {
+        if (start.isNotInRange(4, 8)) {
+            return MGPFallible.failure(MartianChessMoveFailure.START_COORD_OUT_OF_RANGE);
+        }
+        if (end.isNotInRange(4, 8)) {
+            return MGPFallible.failure(MartianChessMoveFailure.END_COORD_OUT_OF_RANGE);
+        }
+        if (end.equals(start)) {
+            return MGPFallible.failure(RulesFailure.MOVE_CANNOT_BE_STATIC());
+        }
+        const dir: MGPFallible<Direction> = Direction.factory.fromDelta(end.x - start.x, end.y - start.y);
+        if (dir.isFailure()) {
+            return MGPFallible.failure(dir.getReason());
+        }
+        return MGPFallible.success(new MartianChessMove(start, end, calledTheClock));
+    }
+    private constructor(start: Coord, end: Coord, public readonly calledTheClock: boolean) {
+        super(start, end);
+    }
+    public toString(): string {
+        const ending: string = this.calledTheClock ? ', CALL_THE_CLOCK' : '';
+        return 'MartianChessMove((' + this.coord.x + ', ' + this.coord.y + ') -> (' +
+                                      this.end.x + ', ' + this.end.y + ')' +
+                                      ending + ')';
+    }
+    public equals(o: MartianChessMove): boolean {
+        if (this.coord.equals(o.coord) === false) return false;
+        if (this.end.equals(o.end) === false) return false;
+        return this.calledTheClock === o.calledTheClock;
+    }
+    public isValidForPawn(): boolean {
+        const dx: number = this.coord.x - this.end.x;
+        const dy: number = this.coord.y - this.end.y;
+        const direction: MGPFallible<Direction> = Direction.factory.fromDelta(dx, dy);
+        return direction.get().isDiagonal();
+    }
+    public isInvalidForDrone(): boolean {
+        const distance: number = this.coord.getOrthogonalDistance(this.end);
+        return distance !== 2;
+    }
+    public isUndoneBy(moveOpt: MGPOptional<MartianChessMove>): boolean {
```

Comment:
ça serait pas plutôt utilisé comme `cancelsMove` ici ? (même si A annule B, A est aussi annulé par B, mais question de clarté)

Reply:
En vrai ni "undo" ni "cancel" ne marche car je crois que c'est vraiment qu'après A->B (call clock ou pas et capture ou pas) tu fait B->A c'est interdit, même si la moindre capture ou clock fait qu'il ne s'agit pas d'une inversion du board.

Ce rappel étant donné, je renomme celle là en isCancelledBy ou cancelsMove ou limite revertsMove ?
Mais oui c'est plus new.cancelsMove(oldMove) c'est vrai !

Reply:
oui vu que c'est pas une annulation mais juste l'opposé (ou l'inverse?), `isInverseOf` pourrait faire l'affaire mais je trouve le `cancelsMove` assez clair, feel free to resolve